### PR TITLE
Staking v4 - scenario nr 2

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request
- New scenario

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: integrationTests/chainSimulator/staking/jail_test.go [hunk 1]
```diff
@@ -152,9 +152,121 @@ func testChainSimulatorJailAndUnJail(t *testing.T, targetEpoch int32, nodeStatus
...
+		MinNodesPerShard:       3,
+		MetaChainMinNodes:      3,
+		AlterConfigsFunction: func(cfg *config.Configs) {
+			cfg.EpochConfig.EnableEpochs.StakingV4Step1EnableEpoch = stakingV4JailUnJailStep1EnableEpoch
+			cfg.EpochConfig.EnableEpochs.StakingV4Step2EnableEpoch = stakingV4JailUnJailStep2EnableEpoch
+			cfg.EpochConfig.EnableEpochs.StakingV4Step3EnableEpoch = stakingV4JailUnJailStep3EnableEpoch
+
+			cfg.EpochConfig.EnableEpochs.MaxNodesChangeEnableEpoch[2].EpochEnable = stakingV4JailUnJailStep3EnableEpoch
+
+			cfg.RatingsConfig.ShardChain.RatingSteps.ConsecutiveMissedBlocksPenalty = 100
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/jail_test.go:192]: do you think it is better to have a function in node/chainSimulator/configs/config.go that will set these values for a very rapid node jailing process?

File: integrationTests/chainSimulator/staking/jail_test.go [hunk 2]
```diff
@@ -152,9 +152,121 @@ func testChainSimulatorJailAndUnJail(t *testing.T, targetEpoch int32, nodeStatus
...
+	txStake := generateTransaction(walletAddress.Bytes, 0, vm.ValidatorSCAddress, minimumStakeValue, txDataField, gasLimitForStakeOperation)
+	stakeTx, err := cs.SendTxAndGenerateBlockTilTxIsExecuted(txStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, stakeTx)
+
+	// wait node to be jailed
+	err = cs.GenerateBlocksUntilEpochIsReached(epochWhenNodeIsJailed)
+	require.Nil(t, err)
+
+	decodedBLSKey1, _ := hex.DecodeString(blsKeys[0])
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/jail_test.go:229]: can we have a direct correspondence between blsKeys[0] and decodedBLSKeyX ? Now we have 0 with 1 and 1 with 2 which makes the test harder to read.

File: integrationTests/chainSimulator/staking/jail_test.go [hunk 3]
```diff
@@ -152,9 +152,121 @@ func testChainSimulatorJailAndUnJail(t *testing.T, targetEpoch int32, nodeStatus
 	checkValidatorStatus(t, cs, blsKeys[0], "eligible")
 }
 
+// Test description
+// Add a new node and wait until the node get jailed
+// Add a second node to take the place of the jailed node
+// UnJail the first node --> should go in queue
+// Activate staking v4 step 1 --> node should be moved from queue to auction list
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/jail_test.go:159]: missing test comment with the scenario number

File: integrationTests/chainSimulator/staking/simpleStake_test.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/multiversx/mx-chain-go/vm"
+	"github.com/stretchr/testify/require"
+)
+
+// Test scenarios
+// Do 3 stake transactions from 3 different wallets - tx value 2499, 2500, 2501
+// testcase1 -- staking v3.5 --> tx1 fail, tx2 - node in queue, tx3 - node in queue with topUp 1
+// testcase2 -- staking v4 step1 --> tx1 fail, tx2 - node in auction, tx3 - node in auction with topUp 1
+// testcase3 -- staking v4 step2 --> tx1 fail, tx2 - node in auction, tx3 - node in auction with topUp 1
+// testcase4 -- staking v3.step3 --> tx1 fail, tx2 - node in auction, tx3 - node in auction with topUp 1
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/simpleStake_test.go:24]: missing test comment with the scenario number

File: integrationTests/chainSimulator/staking/simpleStake_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+		GenesisTimestamp:         startTime,
+		RoundDurationInMillis:    roundDurationInMillis,
+		RoundsPerEpoch:           roundsPerEpoch,
+		ApiInterface:             api.NewNoApiInterface(),
+		MinNodesPerShard:         3,
+		MetaChainMinNodes:        3,
+		NumNodesWaitingListMeta:  3,
+		NumNodesWaitingListShard: 3,
+		AlterConfigsFunction: func(cfg *config.Configs) {
+			cfg.EpochConfig.EnableEpochs.StakingV4Step1EnableEpoch = 2
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/simpleStake_test.go:70]: we might create a "helper" function for these 4 sets by having a "starting" epoch as parameter

File: node/chainSimulator/send_and_execute.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	log.Info("############## send transaction ##############", "txHash", txHashHex)
+
+	return txHashHex, nil
+}
+
+func (s *simulator) SendTxsAndGenerateBlockTilTxIsExecuted(txsToSend []*transaction.Transaction, maxNumOfBlockToGenerateWhenExecutingTx int) ([]*transaction.ApiTransactionResult, error) {
+	hashTxIndex := make(map[string]int)
+	for idx, txToSend := range txsToSend {
```
Comment (iulianpascalau) [inline node/chainSimulator/send_and_execute.go:38]: why not calling here directly the node.GetFacadeHandler().SendBulkTransactions([]*transaction.Transaction{tx}) 
That endpoint will give you the list of hashes, directly

File: node/chainSimulator/send_and_execute.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		txHashHex, err := s.sendTx(txToSend)
+		if err != nil {
+			return nil, err
+		}
+
+		hashTxIndex[txHashHex] = idx
+	}
+
+	time.Sleep(100 * time.Millisecond)
```
Comment (iulianpascalau) [inline node/chainSimulator/send_and_execute.go:47]: do we need this?

File: integrationTests/chainSimulator/staking/jail_test.go (hunk 1 above)
Comment (miiu96) [reply integrationTests/chainSimulator/staking/jail_test.go:192]: added a function

File: integrationTests/chainSimulator/staking/jail_test.go (hunk 2 above)
Comment (miiu96) [reply integrationTests/chainSimulator/staking/jail_test.go:229]: yes, renamed all

File: integrationTests/chainSimulator/staking/jail_test.go (hunk 3 above)
Comment (miiu96) [reply integrationTests/chainSimulator/staking/jail_test.go:159]: added

File: integrationTests/chainSimulator/staking/simpleStake_test.go (hunk 5 above)
Comment (miiu96) [reply integrationTests/chainSimulator/staking/simpleStake_test.go:70]: done

File: node/chainSimulator/send_and_execute.go (hunk 7 above)
Comment (miiu96) [reply node/chainSimulator/send_and_execute.go:47]: No, removed

**Review by iulianpascalau — APPROVED** [review]

**Review by schimih — APPROVED** [review]

**Pull request merged**